applications/DigitalTwinApplication/python_scripts/controls/material_properties_control.py

```diff
@@ -0,0 +1,202 @@
+            if self.controlled_physical_variable not in keys:
+                raise RuntimeError(f"The required gradient for control \"{self.GetName()}\" w.r.t. {self.controlled_physical_variable.Name()} not found. Followings are the variables:\n\t" + "\n\t".join([k.Name() for k in keys]))
+
+            physical_gradient = physical_gradient_variable_container_expression_map[self.controlled_physical_variable]
+            if not IsSameContainerExpression(physical_gradient, self.GetEmptyField()):
+                raise RuntimeError(f"Gradients for the required element container not found for control \"{self.GetName()}\". [ required model part name: {self.primal_model_part.FullName()}, given model part name: {physical_gradient.GetModelPart().FullName()} ]")
+
+            # dj/dE -> physical_gradient
+            # dj/dphi = dj/dphysical * dphysical/dphi
+            return self.filter.BackwardFilterIntegratedField(physical_gradient * self.physical_phi_derivative_field)
+
+    def Update(self, new_control_field: ContainerExpressionTypes) -> bool:
+        if not IsSameContainerExpression(new_control_field, self.GetEmptyField()):
+            raise RuntimeError(f"Updates for the required element container not found for control \"{self.GetName()}\". [ required model part name: {self.primal_model_part.FullName()}, given model part name: {control_field.GetModelPart().FullName()} ]")
+
+        # first clip the control field to max and mins
+        update = new_control_field - self.control_phi_field
+        if Kratos.Expression.Utils.NormL2(update) > 1e-15:
+            with TimeLogger(self.__class__.__name__, f"Updating {self.GetName()}...", f"Finished updating of {self.GetName()}.",False):
+                # update the control thickness field
+                self.control_phi_field = new_control_field
+                # now update the physical field
+                self._UpdateAndOutputFields(update)
+
+                self.filter.Update()
+
+                return True
+        return False
+
+    def _UpdateAndOutputFields(self, update: ContainerExpressionTypes) -> None:
+        # filter the control field
+        filtered_phi_field_update = self.filter.ForwardFilterField(update)
+        self.physical_phi_field = Kratos.Expression.Utils.Collapse(self.physical_phi_field + filtered_phi_field_update)
+
+        # project forward the filtered thickness field
+        physical_field = self.clamper.Clamp(self.physical_phi_field)
+
+        # now update physical field
+        KratosOA.PropertiesVariableExpressionIO.Write(physical_field, self.controlled_physical_variable)
+
+        # compute and store projection derivatives for consistent filtering of the sensitivities
+        # this is dphi/dphysical -> physical_phi_derivative_field
+        self.physical_phi_derivative_field = self.clamper.ClampDerivative(self.physical_phi_field)
```

Comment:
Not a strong point, but I did it to avoid multiple calculations of the same `physical_phi_derivative_field` if algorithm asks within the same iteration.